subsys/emds/emds_flash.c

```diff
@@ -36,6 +38,124 @@ enum ate_type {
 BUILD_ASSERT(offsetof(struct emds_ate, crc8) == sizeof(struct emds_ate) - sizeof(uint8_t),
 	     "crc8 must be the last member");
 
+#define SOC_NV_FLASH_NODE DT_INST(0, soc_nv_flash)
+
+#if NRF52_ERRATA_242_PRESENT
+#include <hal/nrf_power.h>
+/* Disable POFWARN by writing POFCON before a write or erase operation.
+ * Do not attempt to write or erase if EVENTS_POFWARN is already asserted.
+ */
+static bool pofcon_enabled;
+
+static int suspend_pofwarn(void)
+{
+	if (!nrf52_errata_242()) {
+		return 0;
+	}
+
+	bool enabled;
+	nrf_power_pof_thr_t pof_thr;
+
+	pof_thr = nrf_power_pofcon_get(NRF_POWER, &enabled);
+
+	if (enabled) {
+		nrf_power_pofcon_set(NRF_POWER, false, pof_thr);
+
+		/* This check need to be reworked once POFWARN event will be
+		 * served by zephyr.
+		 */
+		if (nrf_power_event_check(NRF_POWER, NRF_POWER_EVENT_POFWARN)) {
+			nrf_power_pofcon_set(NRF_POWER, true, pof_thr);
+			return -ECANCELED;
+		}
+	}
+
+	return 0;
+}
+
+static void restore_pofwarn(void)
+{
+	nrf_power_pof_thr_t pof_thr;
+
+	if (pofcon_enabled) {
+		pof_thr = nrf_power_pofcon_get(NRF_POWER, NULL);
+
+		nrf_power_pofcon_set(NRF_POWER, true, pof_thr);
+		pofcon_enabled = false;
+	}
+}
+
+#define SUSPEND_POFWARN() suspend_pofwarn()
+#define RESUME_POFWARN()  restore_pofwarn()
+#else
+#define SUSPEND_POFWARN() 0
+#define RESUME_POFWARN()
+#endif /* NRF52_ERRATA_242_PRESENT */
+
+static inline bool is_aligned_32(uint32_t data)
+{
+	return (data & 0x3) ? false : true;
+}
+
+static inline bool is_within_bounds(off_t addr, size_t len, off_t boundary_start,
+				    size_t boundary_size)
+{
+	return (addr >= boundary_start &&
+			(addr < (boundary_start + boundary_size)) &&
+			(len <= (boundary_start + boundary_size - addr)));
+}
+
+static inline bool is_regular_addr_valid(off_t addr, size_t len)
+{
+	return is_within_bounds(addr, len, 0, nrfx_nvmc_flash_size_get());
+}
+
+static void nvmc_wait_ready(void)
+{
+	while (!nrfx_nvmc_write_done_check()) {
+	}
+}
+
+static int flash_direct_write(const struct device *dev, off_t offset, const void *data, size_t len)
+{
+	uint32_t flash_addr = offset;
+	uint32_t data_addr = (uint32_t)data;
+
+	ARG_UNUSED(dev);
+
+	if (!is_regular_addr_valid(flash_addr, len)) {
+		return -EINVAL;
+	}
+
+	if (!is_aligned_32(flash_addr)) {
```

Comment:
Who takes care of the alignment?

Reply:
Aligned here: https://github.com/nrfconnect/sdk-nrf/blob/ab06f53b107b202a06e8baa54233d35f3b150a7c/subsys/partition_manager/pm.yml.emds#L7
Maybe the check is not needed, but better to keep it.

Reply:
This is checking the "flash_addr" which is derived from the offset. Though I see that it is incremented with alignment. I was just wondering.